webhook/resources/snapshot/validator.go

```diff
@@ -0,0 +1,91 @@
+package snapshot
+
+import (
+	"fmt"
+	"github.com/longhorn/longhorn-manager/datastore"
+	longhorn "github.com/longhorn/longhorn-manager/k8s/pkg/apis/longhorn/v1beta2"
+	"github.com/longhorn/longhorn-manager/types"
+	"github.com/longhorn/longhorn-manager/webhook/admission"
+	werror "github.com/longhorn/longhorn-manager/webhook/error"
+	admissionregv1 "k8s.io/api/admissionregistration/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"reflect"
+)
+
+type snapshotValidator struct {
+	admission.DefaultValidator
+	ds *datastore.DataStore
+}
+
+func NewValidator(ds *datastore.DataStore) admission.Validator {
+	return &snapshotValidator{ds: ds}
+}
+
+func (o *snapshotValidator) Resource() admission.Resource {
+	return admission.Resource{
+		Name:       "snapshots",
+		Scope:      admissionregv1.NamespacedScope,
+		APIGroup:   longhorn.SchemeGroupVersion.Group,
+		APIVersion: longhorn.SchemeGroupVersion.Version,
+		ObjectType: &longhorn.Snapshot{},
+		OperationTypes: []admissionregv1.OperationType{
+			admissionregv1.Create,
+			admissionregv1.Update,
+		},
+	}
+}
+
+func (o *snapshotValidator) Create(request *admission.Request, newObj runtime.Object) error {
+	snapshot, ok := newObj.(*longhorn.Snapshot)
+	if !ok {
+		return werror.NewInvalidError(fmt.Sprintf("%v is not a *longhorn.Snaphot", newObj), "")
+	}
+
+	if _, ok := snapshot.Labels[types.LonghornLabelVolume]; !ok {
+		return werror.NewInvalidError(fmt.Sprintf("snapshot labels %v doesnot contain %v", snapshot.Labels, types.LonghornLabelVolume), "metadata.labels")
+	}
+
+	if !contains(snapshot.Finalizers, longhornFinalizerKey) {
```

Comment:
Mutator already creates the finalizer, so this is redundant.

Reply:
Thank you. Fixed